static/src/javascripts/projects/journalism/modules/render-campaign.spec.js

```diff
@@ -1,29 +1,81 @@
 // @flow
 
 import { initCampaign } from 'journalism/modules/render-campaign';
-import { getCampaign } from 'journalism/modules/get-campaign';
+import { getCampaigns } from 'journalism/modules/get-campaign';
 
 // hack to stop flow complaining about 'mockImplementation'
 const mock = (mockFn: any) => mockFn;
 jest.mock('journalism/modules/get-campaign');
-mock(getCampaign).mockImplementation(() => ({ title: 'A great campaign' }));
+mock(getCampaigns).mockImplementation(() => [
+    {
+        description: '<p>formatted <b>description</b></p>',
+        formFields: [],
+        formId: 2994970,
+        name: 'callout hij',
+        tagName: 'callout-hij',
+        title: 'Callout HIJ',
+    },
+    {
+        description: '<p>new test<b>&nbsp;</b></p>',
+        formFields: [],
+        formId: 2994970,
+        name: 'callout klm',
+        tagName: 'callout-klm',
+        title: 'Callout klm',
+    },
+]);
 
-describe('Create a campaign element and insert into an article', () => {
-    beforeEach(() => {
-        if (document.body) {
-            document.body.innerHTML =
-                '<div class="content__article-body"><h1>title</h1><p>one</p><p>two</p><p>three</p><p>four</p></div>';
-        }
+const articleBodyWithCalloutBoxes =
+    '<div class="content__article-body"><h1>title</h1><p>one</p><p>two</p><figure data-alt="Callout callout-hij"></figure><p>three</p><p>four</p><figure data-alt="Callout callout-klm"></figure></div>';
+const articleBody =
+    '<div className="content__article-body"><h1>title</h1><p>one</p><p>two</p><p>three</p><p>four</p></div>';
+const renderedCallout =
+    '<figure class="element element-campaign"><div></div></figure>';
+
+const setUpArticle = articleType => {
+    if (document.body) {
+        document.body.innerHTML = articleType;
+    }
+};
+
+const clearUpArticle = () => {
+    if (document.body) {
+        document.body.innerHTML = '';
+    }
+};
+
+describe('If a callout container and callout data are both present', () => {
+    beforeEach(() => setUpArticle(articleBodyWithCalloutBoxes));
+    afterEach(() => clearUpArticle());
+
+    // unfortunately the lodash template used in rendering can't be easily mocked so this basic comparision is all I can do
+    it('puts the callout in the right container', () => {
+        initCampaign();
+        expect(document.querySelectorAll('figure')[0].innerHTML).toEqual(
+            renderedCallout
+        );
```

Comment:
I would like to match this more closely. But I don't know how, because it seems too hard to mock out the lodash template function.. Any ideas welcome

Reply:
Do you mean because of whitespace? yeah this is tricky. I suppose you could write a function:
```js
const nodesEqual = (n1, n2) =>
  n1.nodeType == Node.TEXT_NODE && n1.nodeType == n2.nodeType
    ? n1.trim() == n2.trim()
    : n1.nodeType == Node.ELEMENT_NODE && n1.nodeType == n2.nodeType
    ? attributesEqual(n1.attributes, n2.attributes) && n1.children.every((n, i) => n2.children[i] && nodesEqual(n, n2.children[i]))
    : false;
```
Then you can call it with `nodesEqual(renderedCallout, document.querySelectorAll('figure')[0])`. But that seems overkill :-)

Reply:
Not the whitespace, more the actual html content.  The callout html I use in the test is a poor copy of what it should actually be:
In test - 
```
const renderedCallout =
    '<figure class="element element-campaign"><div></div></figure>';
```
In actual code: 
```
<figure class="element element-campaign"><details class="campaign campaign--snippet">

    <summary class="campaign--snippet__header">
        <div class="campaign--kicker">
            <div class="campaign--snippet__heading-logo">
                <div class="speech-bubble">
                    <h4>Take part</h4>
                </div>
            </div>
            <div class="heading">
                <h4 class="campaign--snippet__headline"> Callout klm</h4>
etc etc ....
```

But we need to simplify somewhere so happy to just leave this